openedx/core/lib/block_cache/transformer.py

```diff
@@ -0,0 +1,123 @@
+"""
+This module provides the abstract base class for all Block Structure
+Transformers.
+"""
+from abc import abstractmethod
+
+
+class BlockStructureTransformer(object):
+    """
+    Abstract base class for all block structure transformers.
+    """
+
+    # All Transformers are expected to maintain a VERSION class
+    # attribute.  While the value for the base class is set to 0,
+    # the value for each concrete transformer should be 1 or higher.
+    #
+    # A transformer's version attribute is used by the block_cache
+    # framework in order to determine whether any collected data for a
+    # transformer is outdated.  When a transformer's data is collected
+    # and cached, it's version number at the time of collection is
+    # stored along with the data.  That version number is then checked
+    # at the time of accessing the collected data (during the transform
+    # phase).
+    #
+    # The version number of a Transformer should be incremented each
+    # time the implementation of its collect method is updated such that
+    # its collected data is changed.
+    #
+    VERSION = 0
+
+    @classmethod
+    def name(cls):
+        """
+        Unique identifier for the transformer's class.  It is used to
+        identify the transformer's cached data. So do not override or
+        modify without consideration of backward compatibility with
+        previously collected data.
+        """
+        return cls.__name__
+
+    @classmethod
+    def collect(cls, block_structure):
+        """
+        Collects and stores any xBlock and modulestore data into the
+        block_structure that's necessary for later execution of the
+        transformer's transform method. Transformers should store such
+        data in the block_structure using the following methods:
+            set_transformer_data
+            set_transformer_block_data
+            request_xblock_fields
+
+        Transformers can call block_structure.request_xblock_fields for
+        any common xBlock fields that should be collected by the
+        framework.
+
+        Any full block tree traversals should be implemented in this
+        collect phase, leaving the transform phase for fast and direct
+        access to a sub-block. If a block's transform output is
+        dependent on its ancestors' data, the ancestor's data should be
+        percolated down to the descendants. So when a (non-root) block
+        is directly accessed in the transform, all of its relevant data
+        is readily available (without needing to access its ancestors).
+
+        Traversals of the block_structure can be implemented using the
+        following methods:
+            topological_traversal
+            post_order_traversal
+
+        Arguments:
+            block_structure (BlockStructureXBlockData) - A mutable
+                block structure that is to be modified with collected
+                data to be cached for the transformer.
+        """
+        pass
```

Comment:
I explicitly chose not to, since I am imagining future transformers that don't have any collection requirements, but just want to transform the blocks.  I believe we have a transformer that does exactly that.